src/lib/server/tmpl_dcursor.c

```diff
@@ -206,6 +206,21 @@ fr_pair_t *_tmpl_cursor_eval(fr_dlist_head_t *list_head, fr_pair_t *curr, tmpl_d
 		break;
 	} else goto all_inst;	/* Used for TMPL_TYPE_LIST */
 
+	/*
+	 *	If no pair was found and there is a fill
+	 *	callback, call that, depending on the suffix
+	 */
+	if ((!vp) && (cc->build) && (ar)) switch (ar->ar_num) {
+	case NUM_UNSPEC:
+	case NUM_LAST:
+	case 0:
```

Comment:
This could perhaps check the `ar->ar_num` against the iteration?  So if there's 2 vps, and the caller says `foo[2] := ...`, it could add the third one.

Tho that is a bit more of an esoteric use-case.

We may also want to create the vp for `NUM_ALL`?  Tho I'd probably start with the 3 items here, and then write test cases to see how we want it to behave.

Reply:
`NUM_ALL` can be added to the above cases.  It's up to the build callback to figure out what it wants to do with it.
I think `case 0` is wrong, it should be checking for `ar->ar_num ==  fr_pair_list_num_elements(&parent->vp_group)`.  Otherwise we'd never be able to add more than one child of a given type to the parent?

Reply:
Is there really much value in having the callback build the pair, seeing as in all cases it's just building a `fr_pair_t` from the `tmpl_attr_t` and appending it to the current list of children?  Seems like we'd just end up with the same boilerplate in every build function.

Reply:
I'd also like to keep this separate from the normal eval function, which is likely to get significantly larger in future when we add proper filter syntax.

Reply:
The `case 0` was partly a work round for tmpl parsing currently not handling TLVs / structs without a suffix.  If we want to add a specific number suffix if all the ones up to that already exist, then it's comparing the iterator count with `ar->ar_num` rather than the number of elements in the parent as those could be any attribute.

Reply:
ah good point